Write to setgroups when doing single user mapping

kpelzel opened the pull request:
There was a [recent commit](https://github.com/containers/buildah/commit/53be3d39f1cb4a182cfc3559726332f643f335c8) to add support for single user mapping without uidmap. Running buildah as an unprivileged user and without having uidmap installed results in this error:

`ERRO[0000] error writing "0 1000 1\n" to /proc/28584/gid_map: write /proc/28584/gid_map: operation not permitted
ERRO[0000] (unable to determine exit status)`

This is due to /proc/[pid]/setgroups not containing 'deny' which is required for unprivileged single user mapping. 

This pull request writes 'deny' to setgroups when doing single user mapping. It also changes the error messages so they are passed through logrus as warnings to prevent an exit status 1 error.

Signed-off-by: Kevin Pelzel <kevinpelzel22@gmail.com>

rhatdan commented on pkg/unshare/unshare.go:
```diff
@@ -195,13 +195,25 @@ func (c *Cmd) Start() error {
 				} else {
-					fmt.Fprintf(continueWrite, "error running newgidmap: %v: %s", err, g.String())
-					fmt.Fprintf(continueWrite, "falling back to single mapping\n")
+					logrus.Warnf("error running newgidmap: %v: %s", err, g.String())
```
I don't think these messages show to the user by default?  Do the messages still show on the terminal?

kpelzel replied:
The logrus.Warnf messages only appear to the user if the --debug flag is on. Is that how you think this should function? When writing to continueWrite before the command would always exit with status code 1.

TomSweeneyRedHat replied:
@nalind thoughts on why the Fprintf to continueWrite would exit out here?

nalind replied:
[The child](https://github.com/containers/buildah/blob/fa67ecb27af4ca6faad5509d9837e05c32ed63cb/pkg/unshare/unshare.c#L229) treats any data that it reads from the reading end of the pipe as an indication that a fatal error has occurred and that it needs to exit.  Messages that aren't about fatal errors should not be written to the pipe.

rhatdan commented on pkg/unshare/unshare.go:
```diff
@@ -195,13 +195,25 @@ func (c *Cmd) Start() error {
+				if c.UseNewgidmap {
+					setgroups, err := os.OpenFile(fmt.Sprintf("/proc/%s/setgroups", pidString), os.O_TRUNC|os.O_WRONLY, 0)
+					if err != nil {
+						fmt.Fprintf(continueWrite, "error opening /proc/%s/setgroups: %v", pidString, err)
```
I believe this message will get displayerd twice.  I don't think you should Fprintf the line.

kpelzel replied:
I was just following the same syntax as lines 219 and 224. Should we remove the Fprintf elsewhere in this function as well?

rhatdan commented on pkg/unshare/unshare.go:
```diff
@@ -195,13 +195,25 @@ func (c *Cmd) Start() error {
+					}
+					defer setgroups.Close()
+					if _, err := fmt.Fprintf(setgroups, "deny"); err != nil {
+						fmt.Fprintf(continueWrite, "error writing 'deny' to /proc/%s/setgroups: %v", pidString, err)
```
I believe this message will get displayerd twice. I don't think you should Fprintf the line.

rhatdan commented:
bot, add author to whitelist
@giuseppe PTAL

giuseppe commented:
LGTM with the changes you've pointed out

giuseppe commented:
please rebase your patch, do not merge master on top of it

kpelzel commented:
Whoops the merge was done on accident. I reset the the head to before the merge and force pushed to remote branch. Then ran a rebase from master.

kpelzel commented:
Related to single-user-mapping issues: https://github.com/containers/buildah/pull/1645

giuseppe commented:
> Whoops the merge was done on accident. I reset the the head to before the merge and force pushed to remote branch. Then ran a rebase from master.

Thanks but I see 3 commits now.  Could you amend two of them and completely drop the merge?

kpelzel commented:
> > Whoops the merge was done on accident. I reset the the head to before the merge and force pushed to remote branch. Then ran a rebase from master.
> 
> Thanks but I see 3 commits now. Could you amend two of them and completely drop the merge?

Ah shoot I messed that up. Would it just make more sense to close this PR in favor of https://github.com/containers/buildah/pull/1645 or would you like me to fix the commits.

giuseppe commented:
> Ah shoot I messed that up. Would it just make more sense to close this PR in favor of #1645 or would you like me to fix the commits.

I am fine if you want to carry the change there, but the two can also be treated separately as this is fixing a bug, while the other PR is adding a new feature

rhatdan commented:
@rh-atomic-bot r+

rhatdan commented:
@rh-atomic-bot retry

rhatdan commented:
@kpelzel This might need to be rebased and repushed. Not sure what failure is being complained about here.

kpelzel commented:
> @kpelzel This might need to be rebased and repushed. Not sure what failure is being complained about here.

Okay I rebased and I think I fixed the weird multiple commits issue.

rhatdan commented:
@rh-atomic-bot r+
